F/kv parser robustness improvements, drop empty keys, better test coverage

bkil-syslogng opened the pull request:
Fixes #1072

bkil-syslogng commented:
Rebased to master. Data race fixes have been cherry picked on top. A leak fix is still expected by tomorrow, which will also be cherry picked on top, so reviewing the previous parts is still valid.

kvch commented on modules/kvformat/tests/test_kv_scanner.c:
```diff
@@ -20,196 +20,302 @@
+    _assert_no_more_tokens(scanner);
 }
 
+#define _INVOKE_VA(args, code) \
```
`_INVOKE_VA(args, code, input)`

kvch commented:
After the small mistake is fixed, this can be merged. :+1:

bkil-syslogng commented:
Fixed comments. At first sight, the mentioned memory leak doesn't apply to this version of kv-parser, so this PR seems final. I'll proceed with investigation on a different track tomorrow, though.

bkil-syslogng commented:
Another `kv-parser` race fixed. The `GString *formatted_key` scratch buffer was also shared. I tried to condition the fixes on whether threading is enabled or not to not cause a speed regression.

bazsi commented on modules/kvformat/kv-parser.c:
```diff
@@ -112,6 +112,28 @@ kv_parser_free(LogPipe *s)
+kv_parser_process_init(LogParser *s, LogMessage **pmsg, const LogPathOptions *path_options, const gchar *input, gsize input_len)
+{
+  if (s->super.cfg && s->super.cfg->threaded)
+    s->process = kv_parser_process_threaded;
```
there are other ways to initiate threading (the threaded flag on an input driver), so this is not really safe.

And I would try to avoid any conditionals in driver code if at all possible.

Can't we avoid malloc/free by using a local variable when cloning? But you know what, only a small fraction of the input will be parsed by kv-parser(), so for now let's stick to the clone() stuff, always. We'll see if there are bottlenecks in the area. If there is, I would probably use a pool of KVScanner instances, one dedicated to all threads.

So I think the conditions on threaded should just be reverted.

bazsi commented:
With the one comment on the conditionals, this could go in.
:+1:

bkil-syslogng commented:
Thank you for your time. I've now removed the check and retained the threaded version only.

While the simple `kv_scanner_clone()` solution looked sane of first glance, unfortunately it would introduce a memory leak which would have been caught by https://github.com/balabit/syslog-ng/pull/1080
I've adjusted it to use `log_pipe_clone()` and `log_pipe_unref()` instead.

Can we now merge this with these two +1's?

[The previous fix proposals used local variables (and sprinkled conditioning) instead. I've opted for this cloning solution now because it is not much slower, while being much easier to understand and more difficult to mess up in the future.]

Backup saved:
* https://github.com/balabit/syslog-ng/compare/master...bkil-syslogng:t/kv-parser-robustness-bk3?expand=1
